### PR TITLE
Core/Opcodes: Implement CMSG_CHANNEL_MODERATE

### PR DESCRIPTION
**Pull request by FALL1N1**
## CHANGES PROPOSED:
-  Implement CMSG_CHANNEL_MODERATE

## ISSUES ADDRESSED:
- None

## TESTS PERFORMED:
- Below

## HOW TO TEST THE CHANGES:
Used mainly for GM Characters, whenever a player sets a password to the world channel or another - you can /join it without having the password and instantly get moderator privileges.
With it - you can also enable/disable the channel moderation.

## KNOWN ISSUES AND TODO LIST:

- None

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/Chat/Channels/Channel.cpp [hunk 1]
```diff
@@ -1260,3 +1260,50 @@ void Channel::RemoveWatching(Player* p)
...
+    }
+
+    // toggle channel moderation
+    _moderation = !_moderation;
+
+    WorldPacket data;
+    if (_moderation)
+        MakeModerationOn(&data, guid);
+    else
+        MakeModerationOff(&data, guid);
```
Comment (FrancescoBorzi) [inline src/server/game/Chat/Channels/Channel.cpp:1294]: guess what... brackets! :D

File: src/server/game/Chat/Channels/Channel.h [hunk 2]
```diff
@@ -224,6 +224,9 @@ class Channel
     void AddWatching(Player* p);
     void RemoveWatching(Player* p);
 
+    // acore
+    void ToggleModeration(Player* p);
```
Comment (FrancescoBorzi) [inline src/server/game/Chat/Channels/Channel.h:228]: why the `acore` comment?

**Comment by FrancescoBorzi** [comment]
nice @FALL1N1 ! just left a couple of minor comments, can be tested already!

File: src/server/game/Chat/Channels/Channel.cpp (hunk 1 above)
Comment (FALL1N1) [reply src/server/game/Chat/Channels/Channel.cpp:1294]: PHP/JS habit :(

File: src/server/game/Chat/Channels/Channel.h (hunk 2 above)
Comment (FALL1N1) [reply src/server/game/Chat/Channels/Channel.h:228]: Thought we'd split the functions in groups for reference purposes, e.g

// trinity

// sunwell

// acore

but it's pointless, so I removed it

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by nicoaravena** [comment]
On a custom channel **could not enter without** a pass and on a global channel could not set  password (I think this is my bad hehehe) 
imagen

The account used to set the password was a normal account and a gm level 3 to enter the channel.

File: src/server/game/Server/WorldSession.h [hunk 3]
```diff
@@ -524,6 +524,7 @@ class WorldSession
     void HandleLootMethodOpcode(WorldPacket& recvPacket);
     void HandleLootRoll(WorldPacket& recvData);
     void HandleRequestPartyMemberStatsOpcode(WorldPacket& recvData);
+    void HandleGroupSwapSubGroupOpcode(WorldPacket& recvData);
```
Comment (pklloveyou) [inline src/server/game/Server/WorldSession.h:527]: Delete here

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/Server/WorldSession.h (hunk 3 above)
Comment (pklloveyou) [reply src/server/game/Server/WorldSession.h:527]: This line of code belongs to another PR

**Comment by Kitzunu** [comment]
@FALL1N1

**Comment by Kitzunu** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/4021/checks?check_run_id=1723029977

**Comment by FALL1N1** [comment]
all good to go

**Comment by Kitzunu** [comment]
Thanks :)

**Pull request merged**